Strengthening three theorems

mazsa opened the pull request:
by weakening their antecedents ( ~ ecelqs , ~ uniqs, and ~ ecelqsdmb ) and making one of them biconditional ( ~ ecelqsdmb ). The rest is Mathbox.

icecream17 commented on set.mm:
```diff
@@ -90762,10 +90796,25 @@ general ordinal versions of these theorems (in this case ~ oa0r ) so
-    uniqs $p |- ( R e. V -> U. ( A /. R ) = ( R " A ) ) $=
+    $( The union of a quotient set.  (Contributed by NM, 9-Dec-2008.)
+       (Proof modification is discouraged.)  (New usage is discouraged.) $)
+    uniqsOLD $p |- ( R e. V -> U. ( A /. R ) = ( R " A ) ) $=
```
similar comment for here
```
uniqsOLD --> uniqs
uniqs    --> uniqsg
```

avekens commented on set.mm:
```diff
@@ -90718,11 +90740,23 @@ general ordinal versions of these theorems (in this case ~ oa0r ) so
-    ecelqsg $p |- ( ( R e. V /\ B e. A ) -> [ B ] R e. ( A /. R ) ) $=
+       Jeff Madsen, 10-Jun-2010.)  (Revised by Mario Carneiro, 9-Jul-2014.)
+       (Proof modification is discouraged.)  (New usage is discouraged.) $)
+    ecelqsOLD $p |- ( ( R e. V /\ B e. A ) -> [ B ] R e. ( A /. R ) ) $=
```
To keep these OLD theorems is not only convenient, but necessary, at least as long as they are used in the proofs of other theorems. The intention of OLD theorem is to be deleted after a year, but this is not possible if they are used by the proofs of not OLD theorems. Moreover, an OLD theorem`s comment must contain "Obsolete version of ~ xxx as of dd-mmm-yyyy, indicating when this theorem can be deleted.

So it is strongly required to respect @icecream17's remarks before this PR can be accepted.

Finally: are there any proofs which can be shortened by the new, weaker versions of the theorems?

mazsa commented:
@avekens @icecream17 I accept the suggestion to relabel the theorems. Implementing a global change throughout the library (including definitional changes) would be too large a task for me. 

However, I would like your advice on two points:

1. Naming and “generalization”.
Mathematically, the new versions only weaken the antecedents; they do not change the conclusions. This makes them generalizations in the usual logical sense, but they are not derivable from the old theorems alone – both old and new are proved from the axioms. Their purpose is to avoid unnecessary hypotheses that would later block definitions and theorems at the optimal level of generality (e.g. statements like ~ erimeq |- ( R e. V -> ( R ErALTV A -> ~ A = R ) ) or ~ mpets |- MembParts = CoMembErs ( ~ mpets2 |- ( A e. V -> ( ( `' _E | A ) Parts A <-> ,~ ( `' _E | A ) Ers A ) ) ).
In our naming conventions, “g” is described as “g : theorem in closed form having an 'is a set' antecedent”, which does fit the new versions syntactically. If “g” is acceptable here, I can use e.g. ecelqsg / uniqsg. If you prefer a different suffix, I am fine with that too. My only concern is to keep the weaker ecelqs / uniqs (in the sense of “w : weak(er) form of a theorem”) intact while having labels for the stronger (= weaker-antecedent) forms that are clearly intended as the ones to use in new developments.

2. Shortening vs. avoiding unnecessary restrictions.
I agree that shortening existing proofs is an important criterion for adding new top-level theorems. However, I don’t think it should be the only one. There are cases where a stronger antecedent is simply non-optimal and excludes statements that are provable from the axioms. For example, with the older version it is not clear how one would obtain ~ cnvepima |- ( A e. V -> ( `' _E " A ) = U. A ) without first switching to a more general form. Similarly, ~ uniqs2 is based on ~ df-er rather than the more general ~ df-ers, so it is unsurprising that it cannot benefit from the weaker antecedent; that is a limitation of the definition, not of the general principle.
My goal here is to set as the default those theorems that do not bake in stronger conditions than necessary and therefore do not silently exclude possibilities that are consistent with our axioms. In many cases this will also shorten proofs; in some cases the main payoff is simply cleaner and more reusable hypotheses. I hope this rationale is acceptable, even when we do not yet have a long list of explicit proof shortenings.

avekens commented:
Since ``( R e. V -> ( R |` A ) e. _V )``  by ~resexg, we can (and should) proof ~ecelqsOLD (however named soon) using ~ecelqs and ~resexg. The same with ~ uniqs(OLD).
By this, the proofs of the current OLD theorems are shortened at least. This would be a sufficient justification to add your new, stronger versions of the theorems in the main body of set.mm.

As for the suffix "g": according to our conventions, the suffix "g" should be used for theorems in closed form derived from inferences "_by replacing the hypothesis with an antecedent of the form ` ( A e. V -> `, we sometimes suffix the closed form with "g" (for "more general") as in ~ uniex versus ~ uniexg._"
Although we do not derive any theorem from an inference here, the result is as it would have been, so we can change the current ~ecelqsOLD into ~ecelqsg (as it was before!), because it has (almost) an antecedent of the form ` ( A e. V -> `. So this would be the exact opposite of @icecream17's suggestion, but I think this would be the best fitting option.

icecream17 commented:
The meaning of "g" I was using is "more general", which is also listed in <https://us.metamath.org/mpeuni/conventions-labels.html> (only in the table, it seems)... but I see it is ambiguous here.

Responding to <https://github.com/metamath/set.mm/pull/5121#issuecomment-3576679350>:

I agree there is an interesting question about choosing "g" + "" vs "" + "w" (weaker version). Personally, I usually just choose the option with the least amount of work and which preserves labelling ("g" for the new theorem + "") unless it is the new theorem is better to work with, i.e. it would be easier on everyone if the new version was the default.

In my opinion, the argument of avoiding unnecessary restrictions (someone who only sees the weaker version is limited) makes sense, but is hopefully reduced by people being able to find the more general version about as easily as finding the normal version. Personally, I wouldn't say that this situation happens enough to be significant (but others disagreeing would change my mind on this), **but** with the ambiguity of "g", using "w" for the weaker version is probably clearer.

mazsa commented:
Changes requested
1. As @avekens suggested, I rederived the weaker theorems. I like his way of keeping the shortness criterion at a primary position, which may have broader significance: If you believe your *theorem* is stronger, you are better served demonstrating it by deriving its weaker versions from it via a much shorter path.

2. Regarding the labels, I implemented @icecream17's latest suggestion: I pasted the *unchanged* labels onto the most general cases, and I added the "w" labels to the weaker ones. The current state is as follows:
The new, logically stronger (weaker antecedent) theorems keep the default ecelqs, uniqs labels.
The old, logically weaker (stronger antecedent) theorems become ecelqsw, uniqsw.

3. Clarified comments of ~ qseq .

mazsa commented:
Oops... it looks like the tags weren't consistent originally (one had a g and the other didn't), so the current version didn't preserve that difference. @avekens please let me know if you find the current state unacceptable:  I'll follow your suggestion exactly tomorrow.

avekens commented:
> Oops... it looks like the tags weren't consistent originally (one had a g and the other didn't), so the current version didn't preserve that difference. @avekens please let me know if you find the current state unacceptable: I'll follow your suggestion exactly tomorrow.

The original naming was inconsistent (in my point of view: also ~uniqs2 should be called ~uniqsd according to our conventions...),
so it is not necessary to keep this difference.

avekens commented:
> The meaning of "g" I was using is "more general", which is also listed in https://us.metamath.org/mpeuni/conventions-labels.html (only in the table, it seems)... but I see it is ambiguous here.
> 
> Responding to [#5121 (comment)](https://github.com/metamath/set.mm/pull/5121#issuecomment-3576679350):
> 
> I agree there is an interesting question about choosing "g" + "" vs "" + "w" (weaker version). Personally, I usually just choose the option with the least amount of work and which preserves labelling ("g" for the new theorem + "") unless it is the new theorem is better to work with, i.e. it would be easier on everyone if the new version was the default.
> 
> In my opinion, the argument of avoiding unnecessary restrictions (someone who only sees the weaker version is limited) makes sense, but is hopefully reduced by people being able to find the more general version about as easily as finding the normal version. Personally, I wouldn't say that this situation happens enough to be significant (but others disagreeing would change my mind on this), **but** with the ambiguity of "g", using "w" for the weaker version is probably clearer.

The meaning of "g" and "w" is explained in bullet point "Suffixes." in https://us.metamath.org/mpeuni/conventions-labels.html:

       <li> g : theorem in closed form having an "is a set" antecedent</li>
       <li> w : weak(er) form of a theorem</li>

Although this specification is very general, I thing the suffix "w" is used for a special purpose, especially in the field of predicate calculus. @wlammen maybe you can tell more about that. 

Beyond propositional calculus, the suffix "w" may be used for other purposes, so I am fine with the current naming.

wlammen commented:
The 'w' suffix is often used for theorems depending on side conditions like implicit substitution ( x = y -> ( ph <-> ps ) ) required by hypothesis.  The particular side condition "disjoint variables" uses a different suffix 'v'.
